Add statistics pages

samuelcambien opened the pull request:
## References
https://github.com/DSpace/Rest7Contract/blob/main/statistics-reports.md
Fixes #731

## Description
This PR implements the statistics pages for DSpace 7.

## Instructions for Reviewers

List of changes in this PR:
* I've added a StatisticsPageModule for all components related to the statistics pages. Added this module to load on the 'statistics' path.
* I've added a generic StatisticsPage component to represent a statistics page.
* I've added subclasses to this component for the site, community, collection and item statistics pages, with the reports to show on those pages. Added corresponding routes to the statistics page routing module.
* I've added a context-sensitive link to the sidebar menu to the relevant statistics page.
* I've added a StatisticsTableComponent to render a usage report as a table.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

tdonohue commented on src/app/+statistics-page/collection-statistics-page/collection-statistics-page.component.ts:
```diff
@@ -0,0 +1,41 @@
+  templateUrl: '../statistics-page/statistics-page.component.html',
+  styleUrls: ['./collection-statistics-page.component.scss']
+})
+export class CollectionStatisticsPageComponent extends StatisticsPageComponent<Collection> {
```
Can we move all these new statistics components/pages under `src/app/statistics-page`?  There's no need for the "+" in the path. See this long standing issue #327

samuelcambien replied:
Ok, renamed the directory

tdonohue commented on src/app/core/submission/usage-report-data.service.ts:
```diff
@@ -0,0 +1,59 @@
+
+@Injectable()
+@dataService(USAGE_REPORT)
+export class UsageReportService extends DataService<UsageReport> {
```
Please add TypeDocs for this new class and its methods

samuelcambien replied:
Added typedocs

tdonohue commented on src/app/core/statistics/models/usage-report.model.ts:
```diff
@@ -0,0 +1,45 @@
+
+@typedObject
+@inheritSerialization(HALResource)
+export class UsageReport extends HALResource {
```
Please add TypeDocs for this new class, and the `Point` interface in this same file

samuelcambien replied:
Added typedocs

tdonohue reviewed (changes requested):
@samuelcambien : I gave this a quick code review today & tested it with the Backend PR (https://github.com/DSpace/DSpace/pull/2782).  Overall, the code looks OK, but I'm not able to get it working as expected (it's possible some of the issues are backend bugs...I'm looking into it).

1.  The site-level statistics page is always empty for me (the page is entirely blank, with just a page heading).  Even after browsing around to items, I cannot get anything to appear at http://localhost:4000/statistics/
2. The "Statistics" menu item (at the top of each page) appears to be caching its path & not updating as I browse from Site to Community to Collection to Item.  If you start at the Site level and browse down, it always remains http://localhost:4000/statistics/   However, if I reload a Collection page in the browser, then that "Statistics" link is updated to the full Collection statistics path.
3. On Community, Collection and Item statistics the results shown are always zero.  This might be a backend bug though, as it appears that is what is returned by my REST API even though I see successful `viewevents` sent each time I browse to a new object.

samuelcambien commented:
Hi Tim,

Thanks for your feedback.
As for the missing statistics (1. and 3.), I cannot reproduce this: with my backend I have statistics on all pages.
I guess these are not front-end issues.
As for 2., the statistics links are not properly updated indeed.

One way to solve this would be
* to remove the 'compareArraysUsingIds()' function here:
https://github.com/DSpace/dspace-angular/blob/main/src/app/shared/menu/menu.component.ts#L82
* and to re-write the menu effects to remove & add the persistent menu sections if they are resolved again, here:
https://github.com/DSpace/dspace-angular/blob/main/src/app/shared/menu/menu.effects.ts#L50

But I think this would decrease performance, so instead I simply renamed the menu id's so they are all unique :)

tdonohue reviewed (changes requested):
@samuelcambien : I've retested this today after I figured out a more [reliable way to generate some `viewevents` statistics in the backend PR](https://github.com/DSpace/DSpace/pull/2782#pullrequestreview-486928465).

I'm **still seeing the following issues:**
1. The site-level statistics page is **still** always empty for me (the page is entirely blank, with just a page heading). I cannot get anything to appear at http://localhost:4000/statistics/   I even tried to generate a `viewevent` on the Site object itself by running this `curl` command: `curl -X POST 'http://localhost:8080/server/api/statistics/viewevents' -H "Content-Type:application/json" -H "X-Forwarded-For:23.249.35.27" --data '{"targetId":"6d65c6a2-3fe7-44dd-bacb-79271257c35d", "targetType":"site"}'`   I still do not see any data on that main page.
2. The "Statistics" menu item (at the top of each page) appears to be caching its path & not updating as I browse from Site to Community to Collection to Item.   _I'm not sure of the best way to solve this issue, but I mentioned it to @artlowel , so hopefully he can give you some advice._

The Community/Collection/Item pages appear to work (after I generated some statistics locally).  I now see report information pertaining to "Total Visits", "Total visits per month", "Top country views" and "Top city views".

As a sidenote, it looks like there's also a small merge conflict in this PR now.

samuelcambien commented:
Hi Tim,
Thanks for your feedback.

* The site statistics page is probably empty for you because it is not supposed to display view events on the site object, instead it shows statistics for all items in the repository, similar to the same [page in DSpace 6](https://demo.dspace.org/xmlui/statistics-home):

I've set up the [backend branch](https://github.com/atmire/dspace/tree/w2p-71214_Usage-reports) locally, but I ran into an authorization issue on the site statistics page.
This is related to [2930](https://github.com/DSpace/DSpace/pull/2930), so I rebased the backend branch on the latest main to have this PR included.
Then I ran the db command mentioned [here](https://github.com/DSpace/DSpace/pull/2930#pullrequestreview-489035503):

Afterwards, I was able to generate some item stats with the request you mentioned (but with uuid of an item).
After committing the solr statistics core (or waiting for 15 minutes) I got data on the site statistics page.

* As for the caching links, I've checked with Art and he agreed that altering the ID's would be the best solution.
I've added an extra fix to append the uuid to the link id, so they would also be unique among items/collections/communities.

* Rebased on the latest master to resolve conflicts.

atarix83 reviewed (changes requested):
thanks @samuelcambien  for this PR
Generally code looks good, I've only added an inline comment.
I've also tested new statistics pages and they seem to work properly.
I've only a comment, it should be nice if the statistic page show a loading spinner while retrieving statistics from server and if any the message `No data available`. What happens now is that the page show immediately the `No data available` message that is replaced by statistics later

samuelcambien commented:
Hi Guiseppe,
Thanks for your feedback.
I moved the data service and also added the ds-loading component.
The 'no data available' message is now only displayed when the reports have loaded.

tdonohue reviewed (approved):
👍 This looks good to me now.  I also ran some tests and it appears to be working. (Note though that I hit bug https://github.com/DSpace/DSpace/issues/2965 again in my testing...so, other testers need to be aware you must have a dbfile installed for stats to be logged).  I expect we'll want to enhance these (very basic) stats reports, but they are a good initial version

atarix83 reviewed (changes requested):
@samuelcambien thanks for addressed my feedback
however making more test, I probably found a bug. Indeed if I show statistics for a community and than trying to show statistics for other community I always get statistics of the first community I've visited. This is valid also between collections and items. Could you check it?

samuelcambien commented:
Hi Guiseppe, thanks for pointing this out.
This was due to a problem in the MenuEffects resolveSubstitutions method, which caused the result to be stored in the route object.
Should be resolved by the latest commit.

atarix83 reviewed (approved):
@samuelcambien thanks for the fix, it works properly now.
@tdonohue I have a suggestion that should be nice to improve in a follow-up PR. 
Basically the breadcrumb in the statistics should include also the link of the related dspace objects, this should allow to easily navigate back to the resource for which statistics are being displayed
